### ISSUE TITLE
[BUG] Fail to upgrade harvester-cloud-provider on RKE2 cluster  (`rancher-v2.7.13`)

### ISSUE THREAD

**Issue by albinsun** [harvester/harvester#5871]
**Describe the bug**
Fail to Upgrade `harvester-cloud-provider` (`rancher-v2.7.13`)

> [!NOTE]
> * Similiar to harvester/harvester/issues/5382

![image](https://github.com/harvester/harvester/assets/2773781/a06eeecb-482b-42d3-96c2-b9ca759b1d24)

**To Reproduce**
1. Setup `harvester-v1.2.2` and `rancher-v2.7.13`
3. Import Harvester to Rancher
4. <details>
   <summary> Create RKE2 cluster (<code>v1.27.13+rke2r1</code>)
   </summary>

   ![image](https://github.com/harvester/harvester/assets/2773781/7a0ce0af-cda3-464f-b393-724f6ff33915)
   `harvester-cloud-provider` and `harvester-csi-driver`
   ![image](https://github.com/harvester/harvester/assets/2773781/7800c3f5-298c-484f-9be9-b822443164c0)
   </details>
5. Upgrade `harvester-cloud-provider` :x: 
   * Upgrade
     ![image](https://github.com/harvester/harvester/assets/2773781/99f10197-24d3-4d39-a51c-448ec26194de)
     ![image](https://github.com/harvester/harvester/assets/2773781/7944efaf-a2d9-4fa4-9846-5c2b6362ce40)
   * After
     ![image](https://github.com/harvester/harvester/assets/2773781/4009f358-5997-405d-a9de-390bad48c92a)
     ![image](https://github.com/harvester/harvester/assets/2773781/df6f5e7a-9019-452b-93a8-808783aad04e)
     ![image](https://github.com/harvester/harvester/assets/2773781/2b4a1b3d-760d-4846-9d21-6fae3851f330)

**Expected behavior**
Should be able to upgrade Rancher harvester app

**Support bundle**
[supportbundle_FailUpgradeCloudProvider_2024-05-24T09-28-52Z.zip](https://github.com/harvester/harvester/files/15431486/supportbundle_FailUpgradeCloudProvider_2024-05-24T09-28-52Z.zip)
[myrke2-pool1-427c78b3-g4hvg-2024-05-24_09_21_21.tar.gz](https://github.com/harvester/harvester/files/15431489/myrke2-pool1-427c78b3-g4hvg-2024-05-24_09_21_21.tar.gz)

**Environment**
* Harvester
  * Version: `v1.2.2`
  * Profile: QEMU/KVM, single nodes (16C/32G/500G)
  * ui-source: `Auto`
* Rancher
  * Version: `v2.7.13`
  * Profile: Helm (`v1.27.13 +k3s1`) in QEMU/KVM (2C/4G)

**Additional context**
1. [helm-operation-htk24_undefined.log](https://github.com/harvester/harvester/files/15431221/helm-operation-htk24_undefined.log)
    ```
    helm upgrade --history-max=5 --install=true --namespace=kube-system --timeout=10m0s --values=/home/shell/helm/values-harvester-cloud-provider-102.0.3-up0.2.4.yaml --version=102.0.3+up0.2.4 --wait=true harvester-cloud-provider /home/shell/helm/harvester-cloud-provider-102.0.3-up0.2.4.tgz
    2024-05-24T09:08:13.917362208Z checking 8 resources for changes
    2024-05-24T09:08:13.939003218Z Looks like there are no changes for ServiceAccount "kube-vip"
    2024-05-24T09:08:13.956356292Z Patch ServiceAccount "harvester-cloud-provider" in namespace kube-system
    2024-05-24T09:08:13.974831362Z Looks like there are no changes for ClusterRole "kube-vip"
    2024-05-24T09:08:13.982302632Z Looks like there are no changes for ClusterRole "harvester-cloud-provider"
    2024-05-24T09:08:13.986955266Z Looks like there are no changes for ClusterRoleBinding "kube-vip"
    2024-05-24T09:08:13.992478676Z Patch ClusterRoleBinding "harvester-cloud-provider" in namespace 
    2024-05-24T09:08:14.004377755Z Patch DaemonSet "kube-vip" in namespace kube-system
    2024-05-24T09:08:14.033197027Z Patch Deployment "harvester-cloud-provider" in namespace kube-system
    2024-05-24T09:08:14.049158884Z beginning wait for 8 resources with timeout of 10m0s
    2024-05-24T09:08:14.097930996Z DaemonSet is not ready: kube-system/kube-vip. 0 out of 1 expected pods have been scheduled
    2024-05-24T09:08:14.124818165Z DaemonSet is not ready: kube-system/kube-vip. 0 out of 1 expected pods have been scheduled
    2024-05-24T09:08:16.151220156Z Deployment is not ready: kube-system/harvester-cloud-provider. 0 out of 1 expected pods are ready
    ...
    2024-05-24T09:18:12.141772503Z Deployment is not ready: kube-system/harvester-cloud-provider. 0 out of 1 expected pods are ready
    2024-05-24T09:18:14.069821673Z Error: UPGRADE FAILED: context deadline exceeded
    ```

2. `rke2-server.service`
    ```
    root@myrke2-pool1-427c78b3-g4hvg:~# journalctl -u rke2-server.service --follow
    May 24 08:31:56 myrke2-pool1-427c78b3-g4hvg rke2[2328]: time="2024-05-24T08:31:56Z" level=error msg="error syncing 'kube-system/rke2-ingress-nginx': handler helm-controller-chart-registration: helmcharts.helm.cattle.io \"rke2-ingress-nginx\" not found, requeuing"
    May 24 08:31:57 myrke2-pool1-427c78b3-g4hvg rke2[2328]: time="2024-05-24T08:31:57Z" level=error msg="error syncing 'kube-system/rke2-metrics-server': handler helm-controller-chart-registration: helmcharts.helm.cattle.io \"rke2-metrics-server\" not found, requeuing"
    May 24 08:31:58 myrke2-pool1-427c78b3-g4hvg rke2[2328]: time="2024-05-24T08:31:58Z" level=error msg="error syncing 'kube-system/rke2-snapshot-controller-crd': handler helm-controller-chart-registration: helmcharts.helm.cattle.io \"rke2-snapshot-controller-crd\" not found, requeuing"
    May 24 08:31:58 myrke2-pool1-427c78b3-g4hvg rke2[2328]: time="2024-05-24T08:31:58Z" level=error msg="error syncing 'kube-system/rke2-snapshot-controller': handler helm-controller-chart-registration: helmcharts.helm.cattle.io \"rke2-snapshot-controller\" not found, requeuing"
    May 24 08:31:59 myrke2-pool1-427c78b3-g4hvg rke2[2328]: time="2024-05-24T08:31:59Z" level=error msg="error syncing 'kube-system/rke2-snapshot-validation-webhook': handler helm-controller-chart-registration: helmcharts.helm.cattle.io \"rke2-snapshot-validation-webhook\" not found, requeuing"
    May 24 08:33:42 myrke2-pool1-427c78b3-g4hvg rke2[2328]: time="2024-05-24T08:33:42Z" level=info msg="Tunnel authorizer set Kubelet Port 10250"
    May 24 08:35:51 myrke2-pool1-427c78b3-g4hvg rke2[2328]: time="2024-05-24T08:35:51Z" level=warning msg="Proxy error: write failed: write tcp 127.0.0.1:9345->127.0.0.1:60592: write: connection reset by peer"
    May 24 08:35:51 myrke2-pool1-427c78b3-g4hvg rke2[2328]: time="2024-05-24T08:35:51Z" level=warning msg="Proxy error: write failed: write tcp 127.0.0.1:9345->127.0.0.1:60594: write: connection reset by peer"
    May 24 08:36:06 myrke2-pool1-427c78b3-g4hvg rke2[2328]: time="2024-05-24T08:36:06Z" level=info msg="Updating TLS secret for kube-system/rke2-serving (count: 11): map[field.cattle.io/projectId:c-m-lqmhzzr9:p-s4z2q listener.cattle.io/cn-10.43.0.1:10.43.0.1 listener.cattle.io/cn-127.0.0.1:127.0.0.1 listener.cattle.io/cn-192.168.0.51:192.168.0.51 listener.cattle.io/cn-__1-f16284:::1 listener.cattle.io/cn-kubernetes:kubernetes listener.cattle.io/cn-kubernetes.default:kubernetes.default listener.cattle.io/cn-kubernetes.default.svc:kubernetes.default.svc listener.cattle.io/cn-kubernetes.default.svc.cluster.local:kubernetes.default.svc.cluster.local listener.cattle.io/cn-localhost:localhost listener.cattle.io/cn-myrke2-pool1-427c78b3-g4hvg:myrke2-pool1-427c78b3-g4hvg listener.cattle.io/fingerprint:SHA1=45D475077B5F9C6CB8A7F2A5FF9C3424E3B4CB41]"
    May 24 08:36:06 myrke2-pool1-427c78b3-g4hvg rke2[2328]: time="2024-05-24T08:36:06Z" level=info msg="Active TLS secret kube-system/rke2-serving (ver=3233) (count 11): map[field.cattle.io/projectId:c-m-lqmhzzr9:p-s4z2q listener.cattle.io/cn-10.43.0.1:10.43.0.1 listener.cattle.io/cn-127.0.0.1:127.0.0.1 listener.cattle.io/cn-192.168.0.51:192.168.0.51 listener.cattle.io/cn-__1-f16284:::1 listener.cattle.io/cn-kubernetes:kubernetes listener.cattle.io/cn-kubernetes.default:kubernetes.default listener.cattle.io/cn-kubernetes.default.svc:kubernetes.default.svc listener.cattle.io/cn-kubernetes.default.svc.cluster.local:kubernetes.default.svc.cluster.local listener.cattle.io/cn-localhost:localhost listener.cattle.io/cn-myrke2-pool1-427c78b3-g4hvg:myrke2-pool1-427c78b3-g4hvg listener.cattle.io/fingerprint:SHA1=45D475077B5F9C6CB8A7F2A5FF9C3424E3B4CB41]"
    ```

3. `rancher-system-agent.service `
    ```
    root@myrke2-pool1-427c78b3-g4hvg:~#  journalctl -u rancher-system-agent.service --follow
    May 24 08:59:05 myrke2-pool1-427c78b3-g4hvg rancher-system-agent[15489]: time="2024-05-24T08:59:05Z" level=info msg="[Applyinator] No image provided, creating empty working directory /var/lib/rancher/agent/work/20240524-085905/a4922eda6712eb2213c9314134fa57d0a0e5f86b2a8adf94f773b2bc708ad70f_0"
    May 24 08:59:05 myrke2-pool1-427c78b3-g4hvg rancher-system-agent[15489]: time="2024-05-24T08:59:05Z" level=info msg="[Applyinator] Running command: sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null]"
    May 24 08:59:05 myrke2-pool1-427c78b3-g4hvg rancher-system-agent[15489]: time="2024-05-24T08:59:05Z" level=info msg="[a4922eda6712eb2213c9314134fa57d0a0e5f86b2a8adf94f773b2bc708ad70f_0:stdout]: Name Location Size Created"
    May 24 08:59:05 myrke2-pool1-427c78b3-g4hvg rancher-system-agent[15489]: time="2024-05-24T08:59:05Z" level=info msg="[Applyinator] Command sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null] finished with err: <nil> and exit code: 0"
    May 24 08:59:05 myrke2-pool1-427c78b3-g4hvg rancher-system-agent[15489]: time="2024-05-24T08:59:05Z" level=info msg="[K8s] updated plan secret fleet-default/myrke2-bootstrap-template-f5pqz-machine-plan with feedback"
    May 24 09:09:06 myrke2-pool1-427c78b3-g4hvg rancher-system-agent[15489]: time="2024-05-24T09:09:06Z" level=info msg="[Applyinator] No image provided, creating empty working directory /var/lib/rancher/agent/work/20240524-090906/a4922eda6712eb2213c9314134fa57d0a0e5f86b2a8adf94f773b2bc708ad70f_0"
    May 24 09:09:06 myrke2-pool1-427c78b3-g4hvg rancher-system-agent[15489]: time="2024-05-24T09:09:06Z" level=info msg="[Applyinator] Running command: sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null]"
    May 24 09:09:06 myrke2-pool1-427c78b3-g4hvg rancher-system-agent[15489]: time="2024-05-24T09:09:06Z" level=info msg="[a4922eda6712eb2213c9314134fa57d0a0e5f86b2a8adf94f773b2bc708ad70f_0:stdout]: Name Location Size Created"
    May 24 09:09:06 myrke2-pool1-427c78b3-g4hvg rancher-system-agent[15489]: time="2024-05-24T09:09:06Z" level=info msg="[Applyinator] Command sh [-c rke2 etcd-snapshot list --etcd-s3=false 2>/dev/null] finished with err: <nil> and exit code: 0"
    May 24 09:09:06 myrke2-pool1-427c78b3-g4hvg rancher-system-agent[15489]: time="2024-05-24T09:09:06Z" level=info msg="[K8s] updated plan secret fleet-default/myrke2-bootstrap-template-f5pqz-machine-plan with feedback"
    ```

**Comment by albinsun** [comment]
Same root cause to #5873
A workaround is to delete the old pod thus new pod can spin up.
Ref. https://github.com/harvester/harvester/issues/5873#issuecomment-2145994679

**Comment by albinsun** [comment]
Test PASS with Rancher Prime `v2.8.6` and `harvester-cloud-provider-v0.2.5+`.

> [!NOTE]
> Note that this issue fixed in `harvester-cloud-provider:v0.2.5` which included in `rancher-v2.8+` [^1], so we use Rancher Prime `v2.8.6` instead of `v2.7.13` reported by ticket.


## Environment
* `harvester-v1.4.0-rc3`
  * Profile: Single node AMD64 QEMU/KVM
  * ui-source: `Auto`
* `rancher-v2.8.6`
  * Profile: Single node AMD64 QEMU/KVM
  * K3s Channel: `v1.28`

## Prerequisite
1. Create image _ubuntu-22.04-jammy-server-cloudimg-amd64.img_
1. Create VM network _mgmt-vlan1_
1. <details>
   <summary> Import Harvester to Rancher
   </summary>
   
   ![image](https://github.com/user-attachments/assets/f9751097-6717-421e-8829-89212e8b5947)
   </details>


## Test Case
### CASE
1. <details>
   <summary> :heavy_check_mark: Create RKE2 cluster (<code>v1.27.16+rke2r2</code>)
   </summary>
   
   ![image](https://github.com/user-attachments/assets/8b82e9bd-e51e-43b3-99ae-657240bc6547)
   </details>

1. <details>
   <summary> :heavy_check_mark: Check <i>harvester-cloud-provider</i> chart version should less than <code>v0.2.5</code>
   </summary>
   
   ![image](https://github.com/user-attachments/assets/645c8918-241b-4a46-876c-fc749191cb91)
   </details>

1. <details>
   <summary> :heavy_check_mark: Deploy <i>mynginx1</i>
   </summary>
   
   ![image](https://github.com/user-attachments/assets/75b5c73b-e496-4f9d-8ceb-e8957698b166)
   </details>

1. <details>
   <summary> :heavy_check_mark: Upgrade <i>harvester-cloud-provider</i> chart to <code>v0.2.5+</code>
   </summary>
   
   * Marketplace
     ![image](https://github.com/user-attachments/assets/1c6a1da4-3857-418f-8149-5c766d7005a0)
   * Upgrade
     ![image](https://github.com/user-attachments/assets/2c11825c-da85-4182-99d7-a8853ea5f932)
   </details>

1. <details>
   <summary> :heavy_check_mark: <i>mynginx1</i> still works,deploy <i>mynginx2</i>
   </summary>
   
   * mynginx1
     ![image](https://github.com/user-attachments/assets/f35027ef-168c-4a5a-866f-0c694f185b2b)
   * mynginx2
     ![image](https://github.com/user-attachments/assets/cbc67f38-612e-4a61-b94d-c0f68f210718)
   </details>


[^1]: https://github.com/rancher/charts/tree/release-v2.8/charts/harvester-cloud-provider
